moshi/src/main/java/com/squareup/moshi/internal/Util.java

```diff
@@ -44,6 +45,19 @@
 public final class Util {
   public static final Set<Annotation> NO_ANNOTATIONS = Collections.emptySet();
   public static final Type[] EMPTY_TYPE_ARRAY = new Type[] {};
+  @Nullable private static final Class<?> DEFAULT_CONSTRUCTOR_MARKER;
+  private static final ConcurrentHashMap<Class<?>, Constructor<?>> DEFAULT_CONSTRUCTOR_CACHE
```

Comment:
What are the implications of storing the default constructor cache in a static field?  Would it be better to cache the default constructor in the generated adapters? That would allow the usage of the cache to be scoped to the usage of the adapters.

If an app does a one time read of a large dataset, to populate a DB or something, then these would be retained for the lifetime of the app, as opposed to being able to be collected once parsing was complete (assuming the adapters were only retained within a specific scope).

Reply:
That's not a bad idea, but storing mutable state in the adapters makes me worried. What would the implementation of that look like? My first thought is basically a volatile field and race construction. The other side in caching it per-adapter is that adapters are cached in moshi, so this would only really work if you're newBuilder-ing moshi instances around and dropping them when they go out of scope

Reply:
That was done in AV: GSON to lazily load child type adapters as needed.  In that case it was due to bad API design, but here it would be a single field. https://github.com/rharter/auto-value-gson/pull/165

This might be a small, unnecessary optimization, but in my cases I use different Moshi instances in different contexts.  

When the app launches I might download new data from the server to update the database depending on the TTL, and those entities aren't read again, so I use a Moshi that just knows about those.

Other places, when I'm writing project files, for instance, my `ProjectRepository` has its' own Moshi instance that only knows about the project related JSON adapters.

Reply:
Yeah I see your point. I think there's a balance that could be had somewhere, but for now I think this approach is fine as a starting point. I modeled it after Butterknife, which does [the same](https://github.com/JakeWharton/butterknife/blob/master/butterknife/src/main/java/butterknife/ButterKnife.java#L83). I figure if there hasn't been memory issues there across its long history, this is probably fine. This implementation defers the reflection/lookup until deserialization-time - so it's fairly lazy, won't occur for serialization, and is likely on a background thread amortized by other I/O going on.

The AV-Gson issue (at least what Adam was trying to solve at uber when he opened the PR) was more about avoiding transitive classloading on initialization.

WDYT?

Reply:
I think the Butterknife case is quite a bit different since that's primarily a terminal dependency, not a transient one.  Since any of your dependencies can use Moshi under the hood, this means you're storing the ref to the constructor of any type that's been deserialized and meets the criteria, no matter where it comes from, and that's unrecoverable memory. 

I'm sure in most cases this isn't a huge deal, but that effectively means that field could theoretically be storing constructor refs for models from all 6 analytics SDKs in your app, plus all 4 ad SDKs, plus that remote logging SDK, remote config SDK, etc. I can't imagine each individual reference is large, but that can add up.

Normally in these situations we can control our memory usage by deciding what we keep references to, but a static field takes away that option.

I'm not looking to die on this hill, and if you don't think it's an issue, and no one else does either, then that's fine with me.

Reply:
Updated in 2ae453d after talking with folks offline